decision_service/dsclientlib/logger/ds_eventhub_client.cc

```diff
@@ -0,0 +1,142 @@
+#include <openssl/hmac.h>
+
+#include "ds_eventhub_client.h"
+
+#include <sstream>
+
+using namespace std::chrono;
+using namespace utility;     // Common utilities like string conversions
+using namespace web;         // Common features like URIs.
+using namespace web::http;   // Common HTTP functionality
+
+namespace decision_service {
+
+	//private helper
+	string_t build_url(const std::string& host, const std::string& name)
+	{
+		//for tests
+		size_t pos = host.find("localhost");
+		if (pos != std::string::npos)
+			return conversions::to_string_t("http://" + host);
+
+		std::ostringstream url;
+		url << "https://" << host << "/" << name << "/messages?timeout=60&api-version=2014-01";
+		auto p = conversions::to_string_t(url.str());
+		return p;
+	}
+  
+	int eventhub_client::send(const std::string& post_data, api_status* status) {
+
+		http_request request(methods::POST);
+		request.headers().add(_XPLATSTR("Authorization"), authorization().c_str());
+		request.headers().add(_XPLATSTR("Host"), _eventhub_host.c_str());
+
+		request.set_body(post_data);
+
+		//TODO fix linux issue if the client is not re-created
+		//web::http::client::http_client client(build_url(_eventhub_host, _eventhub_name));
+		auto request_task = _client.request(request).then([&](http_response response)
+		{
+			//expect http code 201
+			if (response.status_code() == status_codes::Created)
+				return error_code::success;
+
+			//report error
+			if (status)
+			{
+				std::ostringstream error_msg;
+				error_msg << "bad http code (expected 201): " << response.status_code() << std::endl;
+				error_msg << "post_data: " << post_data;
+
+				status->set_error_code(error_code::eventhub_http_bad_status_code);
+				status->set_error_msg(error_msg.str());
+			}
+
+			return error_code::eventhub_http_bad_status_code;
+		});
+
+		int error_code = -1;
+		try
+		{
+			request_task.wait();
+			error_code = request_task.get();
+		}
+		catch (const std::exception& e)
+		{
+			error_code = error_code::eventhub_http_generic;
+
+			//report error
+			if (status)
+			{
+				std::ostringstream error_msg;
+				error_msg << e.what();
+				error_msg << "post_data: " << post_data;
+
+				status->set_error_code(error_code);
+				status->set_error_msg(error_msg.str());
```

Comment:
can we introduce a macro/class/whatever that makes this more elegant? REPORT(status, code) << ".... " << "... ;